hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/index/bloom/BucketizedBloomCheckPartitioner.java

```diff
@@ -75,15 +76,16 @@ public class BucketizedBloomCheckPartitioner extends Partitioner {
    * @param keysPerBucket maximum number of keys to pack in a single bucket
    */
   public BucketizedBloomCheckPartitioner(int targetPartitions, Map<HoodieFileGroupId, Long> fileGroupToComparisons,
-      int keysPerBucket) {
-    this.fileGroupToPartitions = new HashMap<>();
+      int keysPerBucket, boolean isBloomIndexParallelismConfigured) {
+    this.fileGroupToPartitions = new TreeMap<>();
 
-    Map<HoodieFileGroupId, Integer> bucketsPerFileGroup = new HashMap<>();
+    Map<HoodieFileGroupId, Integer> bucketsPerFileGroup = new TreeMap<>();
     // Compute the buckets needed per file group, using simple uniform distribution
     fileGroupToComparisons.forEach((f, c) -> bucketsPerFileGroup.put(f, (int) Math.ceil((c * 1.0) / keysPerBucket)));
     int totalBuckets = bucketsPerFileGroup.values().stream().mapToInt(i -> i).sum();
-    // If totalBuckets > targetPartitions, no need to have extra partitions
-    this.partitions = Math.min(targetPartitions, totalBuckets);
+    // If totalBuckets > targetPartitions, no need to have extra partitions if parallelism is explicitly configured.
+    // if not explicitly configured, we will honor max of (totalBuckets or dynamically derived partition count)
+    this.partitions = isBloomIndexParallelismConfigured ? Math.min(targetPartitions, totalBuckets) : Math.max(targetPartitions, totalBuckets);
```

Comment:
Rename `targetPartitions` to `configuredParallelism`?  Both the docs and code are confusing, e.g., `parallelism is explicitly configured`, does this refer to `targetPartitions`?

Reply:
If not configured, should we use `totalBuckets` directly or have a better way of passing in the `targetPartitions`?